### ISSUE TITLE
jax.lax.scan does not allow static loop variables for computing static values (e.g. for jnp.pad)

### ISSUE THREAD

**Issue by vietluong2110** [jax-ml/jax#27601]
Hello,

I'm encountering an issue with `jax.lax.scan` where the loop variable is always traced rather than a static Python integer. This becomes problematic when I need to compute static values—such as padding configurations for `jnp.pad`—that require plain Python integers.

For example, consider the following code:

```python
def scan_body(res_lo, current_level):
    # current_level is expected to be static, but here it's traced.
    dilation = 2 ** current_level
    padl = dilation * (filt_len // 2 - 1)
    padr = dilation * (filt_len // 2)
    # Compute pad_config for jnp.pad: it requires static integer values.
    pad_config = [(0, 0)] * (res_lo.ndim - 1) + [(padl, padr)]

    res_lo_padded = jnp.pad(res_lo, pad_config, mode="wrap")
    res = jax.lax.conv_general_dilated(
        lhs=res_lo_padded,
        rhs=filt,
        padding=[(0, 0)],
        window_strides=[1],
        rhs_dilation=[dilation],
        dimension_numbers=("NCT", "OIT", "NCT"),
        precision=jax.lax.Precision(precision),
    )
    res_lo_new, res_hi = jnp.split(res, 2, axis=1)
    return res_lo_new, res_hi.squeeze(1)

# Use lax.scan with a traced sequence.
final_res_lo, details = jax.lax.scan(scan_body, data, np.arange(level))
 ```
Is there a way to implement scan so that the scan body function receives a static iteration index (as a plain Python integer) rather than a traced value (current_level in this case) ?

Thank you for your help,
Viet Luong

**Comment by superbobry** [comment]
Could you use a plain `for` loop? Doing so will will increase the size of the staged out computation (since the loop will effectively get unrolled), but it does allow you to vary the padding configuration per iteration. 

```
for current_level in range(level):
  # scan body goes here
```

**Comment by vietluong2110** [comment]
But it will also increase the gpu memory allocation. I want to optimize the gpu memory allocation

**Comment by froystig** [comment]
Why would GPU memory allocation increase? Are you seeing this measured in an example? Generally speaking, "unrolling" loops can affect what compiler optimizations take place, but sometimes for the better.

**Comment by vietluong2110** [comment]
@froystig I have not said the for loop increasing the gpu memory allocation. I just have the for loop version and want see whether I can reduce the gpu memory allocation more by using scan instead.

**Comment by froystig** [comment]
`scan` reflects a staged out, rolled loop. By the time loop iterations execute, there are no "static" values to reflect on.

Specifically considering your example, it's not clear how to adapt it to use `scan`. A rolled loop in our backend compiler language (XLA HLO) requires a single body computation – i.e. one that is uniform across iterations. Meanwhile, two `conv_general_dilated` calls with different dilation parameters result in two distinct computations in this language.
